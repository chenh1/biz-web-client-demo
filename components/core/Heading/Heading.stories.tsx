import type { HeadingType } from "./Heading";
import React from 'react';
import { Heading } from './Heading';

type Args = {
  args: HeadingType
}

const Template = (args: HeadingType) => <Heading {...args}>Heading</Heading>;
export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const H1: Args = {
  args: {
    type: "h1"
  },
};

export const H2: Args = {
  args: {
    type: "h2"
  },
};

export const H3: Args = {
  args: {
    type: "h3"
  },
};