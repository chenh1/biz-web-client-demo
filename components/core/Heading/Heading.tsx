import type { CoreColors } from '../../constants/colors'
import React, { useMemo, ReactNode } from 'react';

type HeaderType = 'h1' | 'h2' | 'h3' | 'h4' | `sm:text-${string}` | `md:text-${string}`
type FontSizeClass = string
export type HeadingType = { 
  type: HeaderType, 
  fontSizeClass?: FontSizeClass, 
  children?: string | JSX.Element | JSX.Element[] | ReactNode, 
  className?: string, 
  colorClass?: `text-${CoreColors}`, 
  darkColorClass?: `dark:text-${CoreColors}`
}

const getHeading = (type: HeaderType): string | ((arg0: any) => JSX.Element) => {
  switch(type) {
    case 'h1': {
      return (props) => <h1 {...props}/>;
    }
    case 'h2': {
      return (props) => <h2 {...props}/>;
    }
    case 'h3': {
      return (props) => <h3 {...props}/>;
    }
  }

  return 'h4'
}

const getSizeClasses = (type: HeaderType, fontSizeClass: FontSizeClass): string => {
  if (fontSizeClass) {
    return fontSizeClass;
  }

  switch(type) {
    case 'h1': {
      return 'text-4xl sm:text-6xl'
    }
    case 'h2': {
      return 'text-3xl sm:text-5xl'
    }
    case 'h3': {
      return 'text-2xl sm:text-4xl'
    }
    case 'h4': {
      return 'text-lg sm:text-md'
    }
  }

  if (typeof(type) !== 'undefined') {
    return type
  }

  return 'text-xl sm:text-2xl'
}

const BASE_HEADING_CLASSES: string = 'font-headings font-semibold';

export const Heading: React.FC<HeadingType> = ({ type, fontSizeClass = "", children, className = "", colorClass="text-night-black", darkColorClass="dark:text-max-white", ...rest }) => {
  const H = useMemo(() => {
    return getHeading(type)
  }, [type])

  const computedClasses = useMemo(() => {
    return getSizeClasses(type, fontSizeClass);
  }, [type, fontSizeClass]);

  return (
    <H className={`${colorClass} ${darkColorClass} ${BASE_HEADING_CLASSES} ${computedClasses} ${className}`} {...rest}>
      {children}
    </H>
  )
};