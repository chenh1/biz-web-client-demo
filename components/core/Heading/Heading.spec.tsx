import { Heading } from "./Heading";
import { render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders an h1 tag', async () => {
  const { asFragment } = render(
    <Heading type="h1"/>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders an h1 with fontSizeClass overriding base h1 styles', async () => {
  const { asFragment } = render(
    <Heading type="h1" fontSizeClass="some-class"/>
  );

  expect(asFragment()).toMatchSnapshot();
});
