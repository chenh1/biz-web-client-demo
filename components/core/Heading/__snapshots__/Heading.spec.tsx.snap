// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an h1 tag 1`] = `
<DocumentFragment>
  <h1
    class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl "
  />
</DocumentFragment>
`;

exports[`renders an h1 with fontSizeClass overriding base h1 styles 1`] = `
<DocumentFragment>
  <h1
    class="text-night-black dark:text-max-white font-headings font-semibold some-class "
  />
</DocumentFragment>
`;
